Kerberos.NET/Client/Krb5CredentialCache.cs

```diff
@@ -0,0 +1,600 @@
+            {
+                WriteData(Encoding.UTF8.GetBytes(name), buffer);
+            }
+        }
+
+        private static PrincipalName ReadPrincipal(NdrBuffer buffer)
+        {
+            var type = (PrincipalNameType)buffer.ReadInt32BigEndian();
+            var count = buffer.ReadInt32BigEndian();
+
+            var realmData = ReadData(buffer);
+
+            string realm = "";
+
+            if (realmData.length > 0)
+            {
+                realm = Encoding.UTF8.GetString(realmData.value.ToArray());
+            }
+
+            var components = new List<string>();
+
+            for (var i = 0; i < count; i++)
+            {
+                var component = ReadData(buffer);
+
+                components.Add(Encoding.UTF8.GetString(component.value.ToArray()));
+            }
+
+            return new PrincipalName(type, realm, components);
+        }
+
+        private static void WriteData(ReadOnlyMemory<byte> data, NdrBuffer buffer)
+        {
+            buffer.WriteInt32BigEndian(data.Length);
+            buffer.WriteMemory(data);
+        }
+
+        private static (int length, ReadOnlyMemory<byte> value) ReadData(NdrBuffer buffer)
+        {
+            /*
+            * data ::=
+            *     length (32 bits)
+            *     value (length bytes)
+             */
+
+            var length = buffer.ReadInt32BigEndian();
+            var value = buffer.ReadMemory(length);
+
+            return (length, value);
+        }
+
+        private void WriteHeader(NdrBuffer buffer)
+        {
+            var headerBuffer = new NdrBuffer(align: false);
+
+            foreach (var kv in this.Header)
+            {
+                headerBuffer.WriteInt16BigEndian((short)kv.Key);
+                headerBuffer.WriteInt16BigEndian((short)kv.Value.Length);
+                headerBuffer.WriteMemory(kv.Value);
+            }
+
+            var header = headerBuffer.ToMemory(0);
+
+            buffer.WriteInt16BigEndian((short)header.Length);
+            buffer.WriteMemory(header);
+        }
+
+        private void ReadHeader(NdrBuffer buffer)
+        {
+            var headerLength = buffer.ReadInt16BigEndian();
+
+            int headerRead = 0;
+
+            do
+            {
+                var tag = (Krb5CredentialCacheTag)buffer.ReadInt16BigEndian();
+                var length = buffer.ReadInt16BigEndian();
+                var value = buffer.ReadMemory(length);
+
+                this.Header[tag] = value;
+
+                headerRead += 4 + length;
+            }
+            while (headerRead < headerLength);
+        }
+
+        private static int GetEpoch(DateTimeOffset dt)
+        {
+            TimeSpan ts;
+
+            if (dt == DateTimeOffset.MinValue)
+            {
+                ts = TimeSpan.Zero;
+            }
+            else
+            {
+                ts = dt.Subtract(UNIX_EPOCH_BASE);
+            }
+
+            return (int)ts.TotalSeconds;
+        }
```

Comment:
```suggestion
        private static int GetEpoch(DateTimeOffset dt) => (int)dt.ToUnixTimeSeconds();
```

Reply:
Still needed the if (dt == MinValue) check, but nice catch. Forgot about From/ToUnix